### PR TITLE
planner: range columns partitioning with explicit collation fix (#32749)

### LINKED ISSUE #32749: partition by RANGE COLUMNS on varchar column with collate _bin does not adjust pruning if query includes other collation
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t (a varchar(255) charset utf8mb4 collate utf8mb4_bin) partition by range columns (a) (partition puppera values less than ("AAA"), partition plowera values less than ("aaaa"));
insert into t values ("aa");
explain select * from t where a = "aa" collate utf8mb4_general_ci;
select * from t where a = "aa" collate utf8mb4_general_ci;
```

### 2. What did you expect to see? (Required)

Skip pruning (or have some way to take into consideration the given collation), like MySQL:
```
mysql> create table t (a varchar(255) charset utf8mb4 collate utf8mb4_bin) partition by range columns (a) (partition puppera values less than ("AAA"), partition plowera values less than ("aaaa"));
Query OK, 0 rows affected (0,02 sec)

mysql> insert into t values ("aa");
Query OK, 1 row affected (0,00 sec)

mysql> explain select * from t where a = "aa" collate utf8mb4_general_ci;
+----+-------------+-------+-----------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions      | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+-----------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t     | puppera,plowera | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+-----------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0,00 sec)

mysql> select * from t where a = "aa" collate utf8mb4_general_ci;
+------+
| a    |
+------+
| aa   |
+------+
1 row in set (0,00 sec)
```

### 3. What did you see instead (Required)
Using the same pruning as if the `collate` was not given in the select query:
```
tidb> create table t (a varchar(255) charset utf8mb4 collate utf8mb4_bin) partition by range columns (a) (partition puppera values less than ("AAA"), partition plowera values less than ("aaaa"));
Query OK, 0 rows affected (0,00 sec)

tidb> insert into t values ("aa");
Query OK, 1 row affected (0,00 sec)

tidb> explain select * from t where a = "aa" collate utf8mb4_general_ci;
+-------------------------+----------+-----------+----------------------------+--------------------------------+
| id                      | estRows  | task      | access object              | operator info                  |
+-------------------------+----------+-----------+----------------------------+--------------------------------+
| TableReader_8           | 8000.00  | root      |                            | data:Selection_7               |
| └─Selection_7           | 8000.00  | cop[tikv] |                            | eq(test.t.a, "aa")             |
|   └─TableFullScan_6     | 10000.00 | cop[tikv] | table:t, partition:puppera | keep order:false, stats:pseudo |
+-------------------------+----------+-----------+----------------------------+--------------------------------+
3 rows in set (0,00 sec)

tidb> select * from t where a = "aa" collate utf8mb4_general_ci;
Empty set (0,00 sec)
```

### 4. What is your TiDB version? (Required)

```
tidb_version(): Release Version: v5.5.0-alpha-442-gfbfe36ca7d
Edition: Community
Git Commit Hash: fbfe36ca7dd3a78a714872d468cebfce34e50116
Git Branch: master
UTC Build Time: 2022-03-02 11:21:12
GoVersion: go1.17
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: close #32749

Problem Summary:
When using explicit collation different from the partitioning column collation, the pruning is not correct.

### What is changed and how it works?
If the partition column's collation is binary, then we fall back to use all partitions,
if it is non-binary but the condition is explicitly binary collation then we can pretend it is matching the column's collation (since it is 1:N match and any N is OK as mapping).

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
For PARTITION BY RANGE COLUMNS (char_based_col) and condition with different explicit collation, we now loosen the pruning to match all partitions if explicit non-binary collation.
```

### CODE REVIEW

**Comment by mjonss** [comment]
Two other solutions may work:
1) brute force test; generate all combinations of the possible binary strings matching the non-binary collation, like condition `= "aa"` => "aa", "aA", "Aa", "AA", "aå", "aä" etc. This will be too expensive, especially when considering emojis etc which all are considered matching in _ci collations
2) Limited brute force test; Only generate different matching first rune and prune on those different values, like `= "a"` => "a", "A", "å", "Å", "ä", "Ä" etc. This would at least limit the performance cost, but still I think it is too complex to implement and maintain for a relative little gain.

Current proposal makes the result correct, and is in-line with MySQL's pruning.

File: planner/core/rule_partition_processor.go [hunk 1]
```diff
@@ -1575,13 +1576,32 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
...
+		colCharSet, colColl := p.partCol.RetType.Charset, p.partCol.RetType.Collate
+		if !collate.IsBinCollation(conColl) || conCharSet != colCharSet {
+			return 0, len(p.data), true
+		}
+		clonedCon := con.Clone()
+		if newCon, ok := clonedCon.(*expression.Constant); ok {
+			con = newCon
+		}
+		con.SetCharsetAndCollation(conCharSet, colColl)
+	}
```
Comment (xiongjiwei) [inline planner/core/rule_partition_processor.go:1596]: I think we can fall back to all partitions if the collation used in the expression is not the same as the partition column collation.

And for the `_bin` collation, we can only consider the `eq` expression. MySQL does not prune anything if not `eq`
```sql
mysql> create table t (a varchar(255) charset utf8mb4 collate utf8mb4_general_ci) partition by range columns (a) (partition puppera values less than ("AAA"), partition plowera values less than ("aaaa"), partition pp values less than ('hhhh'));
Query OK, 0 rows affected (0.04 sec)

mysql> insert into t values ("aa");
Query OK, 1 row affected (0.02 sec)

mysql> explain select * from t where a > "AAAa" collate utf8mb4_bin;
+----+-------------+-------+--------------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions         | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+--------------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t     | puppera,plowera,pp | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+--------------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.00 sec)

mysql> explain select * from t where a = "AAAa" collate utf8mb4_bin;
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | t     | pp         | ALL  | NULL          | NULL | NULL    | NULL |    1 |   100.00 | Using where |
+----+-------------+-------+------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.00 sec)
```

Comment (mjonss) [reply planner/core/rule_partition_processor.go]: Thank you for finding this issue! I added tests and the restriction to only support `eq` if explicit _bin collation which is different from partition column/definition.

File: planner/core/rule_partition_processor.go [hunk 2]
```diff
@@ -1575,13 +1576,32 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
 		return 0, len(p.data), false
 	}
 
+	exprCollation, err := expression.CheckAndDeriveCollationFromExprs(sctx, opName, col.RetType.EvalType(), con, p.data[0])
```
Comment (xiongjiwei) [inline planner/core/rule_partition_processor.go:1579]: I think this part can be simplified to
```go
	_, collation := op.CharsetAndCollation()
	if collation != p.partCol.RetType.Collate && !(collate.IsBinCollation(collation) && opName == ast.EQ) {
		return 0, len(p.data), false
	}
```

only focus on the collation of expression and partition column, no matter what the coercibility is.

File: planner/core/rule_partition_processor.go [hunk 3]
```diff
@@ -1575,13 +1576,32 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
...
+		con.SetCharsetAndCollation(conCharSet, colColl)
+	}
 	start, end := p.pruneUseBinarySearch(sctx, opName, con, op)
 	return start, end, true
 }
 
 func (p *rangeColumnsPruner) pruneUseBinarySearch(sctx sessionctx.Context, op string, data *expression.Constant, f *expression.ScalarFunction) (start int, end int) {
 	var err error
 	var isNull bool
+	charSet, collation := p.partCol.RetType.Charset, p.partCol.RetType.Collate
```
Comment (xiongjiwei) [inline planner/core/rule_partition_processor.go:1604]: the parameter `f` is not needed anymore after this line

Comment (mjonss) [reply planner/core/rule_partition_processor.go:1603]: Thanks, fixed

File: planner/core/rule_partition_processor.go (hunk 2 above)
Comment (mjonss) [reply planner/core/rule_partition_processor.go]: Thank you for the suggestion, fixed it and it also improved the pruning when the collation was explicitly set to the same as the column :)

File: planner/core/rule_partition_processor.go [hunk 4]
```diff
@@ -1575,13 +1576,31 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
...
 
-	start, end := p.pruneUseBinarySearch(sctx, opName, con, op)
+	// If different character set or collation, we can only prune if:
+	// - constant is binary collation (can only be found in one partition)
+	// - same char set
+	// - EQ operator, consider values 'a','b','ä' where 'ä' would be in the same partition as 'a' if general_ci, but is binary after 'b'
+	// otherwise return all partitions / no pruning
+	conCharSet, conColl := con.CharsetAndCollation()
+	colCharSet, colColl := p.partCol.RetType.Charset, p.partCol.RetType.Collate
+	if (conCharSet != "" && conCharSet != colCharSet) || (conColl != "" && conColl != colColl) {
```
Comment (xiongjiwei) [inline planner/core/rule_partition_processor.go:1586]: do we have a test case like
```sql
mysql> create table t (a varchar(255) charset utf8mb4 collate utf8mb4_bin) partition by range columns (a) (partition puppera values less than ("AAA"), partition plowera values less than ("aaaa"));
Query OK, 0 rows affected (0,02 sec)

mysql> insert into t values ("aa");
Query OK, 1 row affected (0,00 sec)

mysql> explain select * from t where a collate utf8mb4_general_ci = "aa";
```
it should be the same result with `explain select * from t where a = "aa" collate utf8mb4_general_ci;`

File: planner/core/rule_partition_processor.go [hunk 5]
```diff
@@ -1575,13 +1576,31 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
...
+		if !collate.IsBinCollation(conColl) || conCharSet != colCharSet || opName != ast.EQ {
+			return 0, len(p.data), true
+		}
+		clonedCon := con.Clone()
+		if newCon, ok := clonedCon.(*expression.Constant); ok {
```
[the remark this answers is not in the document]
Comment (mjonss) [reply planner/core/rule_partition_processor.go]: Thanks, fixed in a new commit

File: planner/core/rule_partition_processor.go [hunk 6]
```diff
@@ -1575,13 +1576,31 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
 		return 0, len(p.data), false
 	}
 
-	start, end := p.pruneUseBinarySearch(sctx, opName, con, op)
+	// If different character set or collation, we can only prune if:
+	// - constant is binary collation (can only be found in one partition)
+	// - same char set
+	// - EQ operator, consider values 'a','b','ä' where 'ä' would be in the same partition as 'a' if general_ci, but is binary after 'b'
+	// otherwise return all partitions / no pruning
+	conCharSet, conColl := con.CharsetAndCollation()
```
[the remark this answers is not in the document]
Comment (mjonss) [reply planner/core/rule_partition_processor.go]: @xiongjiwei PTAL on the new commit, if I got the charset and collation correct this time?

File: planner/core/rule_partition_processor.go (hunk 4 above)
Comment (mjonss) [reply planner/core/rule_partition_processor.go]: No, there were no such test, so I added it.

File: planner/core/rule_partition_processor.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply planner/core/rule_partition_processor.go]: yes, that is totally right.

File: planner/core/rule_partition_processor.go [hunk 7]
```diff
@@ -1575,13 +1576,29 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
...
+	// otherwise return all partitions / no pruning
+	_, exprColl := expr.CharsetAndCollation()
+	colColl := p.partCol.RetType.Collate
+	if exprColl != "" && exprColl != colColl {
+		if !collate.IsBinCollation(exprColl) || opName != ast.EQ {
+			return 0, len(p.data), true
+		}
+		con = con.Clone().(*expression.Constant)
+		conCharSet, _ := con.CharsetAndCollation()
+		con.SetCharsetAndCollation(conCharSet, colColl)
```
Comment (xiongjiwei) [inline planner/core/rule_partition_processor.go:1592]: I am a little confusing why we need these 3 lines

**Review by xiongjiwei — APPROVED** [review]

File: planner/core/rule_partition_processor.go [hunk 8]
```diff
@@ -1575,13 +1576,23 @@ func (p *rangeColumnsPruner) partitionRangeForExpr(sctx sessionctx.Context, expr
...
 	}
 
-	start, end := p.pruneUseBinarySearch(sctx, opName, con, op)
+	// If different collation, we can only prune if:
+	// - expression is binary collation (can only be found in one partition)
+	// - EQ operator, consider values 'a','b','ä' where 'ä' would be in the same partition as 'a' if general_ci, but is binary after 'b'
+	// otherwise return all partitions / no pruning
+	_, exprColl := expr.CharsetAndCollation()
+	colColl := p.partCol.RetType.Collate
+	if exprColl != "" && exprColl != colColl && (opName != ast.EQ || !collate.IsBinCollation(exprColl)) {
```
Comment (xiongjiwei) [inline planner/core/rule_partition_processor.go:1585]: I think we can remove the condition `exprColl != ""`, even if the `exprColl` could be `""`, it should be 
```suggestion
	if exprColl == "" || exprColl != colColl && (opName != ast.EQ || !collate.IsBinCollation(exprColl)) {
```

File: ddl/sanity_check.go [hunk 9]
```diff
@@ -92,7 +92,7 @@ func (d *ddl) checkDeleteRangeCnt(job *model.Job) {
 		var physicalTableIDs []int64
 		var ruleIDs []string
 		if err := job.DecodeArgs(&startKey, &physicalTableIDs, &ruleIDs); err != nil {
-			panic("should not happened")
+			panic("Error in drop/truncate table, please report a bug with this stack trace and how it happened")
```
Comment (bb7133) [inline ddl/sanity_check.go:95]: Good point.

**Review by bb7133 — APPROVED** [review]
LGTM

**Pull request merged**